src/main/scala/chisel3/UnitTest.scala

```diff
@@ -0,0 +1,124 @@
+import java.util.jar.JarFile
+import scala.collection.JavaConverters._
+
+/** All classes and objects marked as `UnitTest` are automatically discoverable
+ *  by the `DiscoverUnitTests` helper. */
+trait UnitTest
+
+/** Helper to discover all subtypes of `UnitTest` in the class path, and call
+ *  their constructors (if they are a class) or ensure that the singleton is
+ *  constructed (if they are an object).
+ *
+ *  This code is loosely based on the test suite discovery in scalatest, which
+ *  performs the same scan over the classpath JAR files and directories, and
+ *  guesses class names based on the encountered directory structure. */
+object DiscoverUnitTests {
+
+  /** The callback invoked for each unit test class name and unit test
+   *  constructor. */
+  type Callback = (String, () => Unit) => Unit
+
+  /** Discover all tests in the classpath and call `cb` for each. */
+  def apply(cb: Callback): Unit = classpath().foreach(discoverFile(_, cb))
+
+  /** Return the a sequence of files or directories on the classpath. */
+  private def classpath(): Iterable[File] = System
+    .getProperty("java.class.path")
+    .split(File.pathSeparator)
+    .map(s => if (s.trim.length == 0) "." else s)
+    .map(new File(_))
+
+  /** Discover all tests in a given file. If this is a JAR file, looks through
+   *  its contents and tries to find its classes. */
+  private def discoverFile(file: File, cb: Callback): Unit = {
+    if (file.getPath.toLowerCase.endsWith(".jar")) {
+      val jarFile = new java.util.jar.JarFile(file)
+      jarFile.entries.asScala.foreach { jarEntry =>
+        val name = jarEntry.getName
+        if (!jarEntry.isDirectory && name.endsWith(".class"))
+          discoverClass(pathToClassName(name), cb)
+      }
+    } else if (file.isDirectory) {
+      def visit(prefix: String, file: File): Unit = {
+        val name = prefix + "/" + file.getName
+        if (file.isDirectory) {
+          for (entry <- file.listFiles)
+            visit(name, entry)
+        } else if (name.endsWith(".class")) {
+          discoverClass(pathToClassName(name), cb)
+        }
+      }
+      for (entry <- file.listFiles)
+        visit("", entry)
+    }
+  }
+
+  /** Convert a file path to a class */
+  private def pathToClassName(path: String): String =
+    path.replace('/', '.').replace('\\', '.').stripPrefix(".").stripSuffix(".class")
+
+  /** Load the given class and check whether it is a subtype of `UnitTest`. If
+   *  it is, call its constructor if it is a class, or ensure it is constructed
+   *  if it is an object. */
+  private def discoverClass(className: String, cb: Callback): Unit = {
+    val clazz =
+      try {
+        classOf[UnitTest].getClassLoader.loadClass(className)
+      } catch {
+        case e: ClassNotFoundException       => return
+        case e: NoClassDefFoundError         => return
+        case e: ClassCastException           => return
+        case e: UnsupportedClassVersionError => return
```

Comment:
Consider:

```suggestion
        case _: ClassNotFoundException | _: NoClassDefFoundError | _: ClassCastException | _: UnsupportedClassVersionError => return
```

This sometimes gets _worse_ when you are matching on a large number of types. See what you think.

Also, at minimum `e` is captured, but not used, so use of `_` is better.

Reply:
:+1: Good point about the `e` never being used. Collapsing stuff on one line feels a bit less legible.